drivers/ethernet/intel/eth_intel_igc.c

```diff
@@ -0,0 +1,1349 @@
+	static void eth##n##_cfg_irq(const struct device *dev)                                     \
+	{                                                                                          \
+		eth_intel_igc_init_queue_work_##n(dev);                                            \
+	}
+
+/**
+ * @brief Allocate various global objects required for managing tx and rx operations.
+ */
+#define INTEL_IGC_ALLOC_GLOBAL_OBJECTS(n)                                                          \
+	struct k_sem tx_ring_lock_##n[NUM_QUEUES(n)];                                              \
+	struct k_sem rx_ring_lock_##n[NUM_QUEUES(n)];                                              \
+	static unsigned int tx_ring_wr_ptr_##n[NUM_QUEUES(n)];                                     \
+	static unsigned int rx_ring_wr_ptr_##n[NUM_QUEUES(n)];                                     \
+	static unsigned int tx_ring_rd_ptr_##n[NUM_QUEUES(n)];                                     \
+	static unsigned int rx_ring_rd_ptr_##n[NUM_QUEUES(n)];                                     \
+	static struct net_buf tx_frag_##n[NUM_QUEUES(n)][NUM_TX_DESC(n)];                          \
+	static struct net_pkt tx_pkt_##n[NUM_QUEUES(n)][NUM_TX_DESC(n)];                           \
+	static struct eth_intel_igc_intr_info intr_info_##n[NUM_MSIX(n)];                          \
+	static msi_vector_t msi_vec_##n[NUM_MSIX(n)];
+
+/**
+ * @brief Initializes the runtime data structures of each driver instance.
+ */
+#define INTEL_IGC_MAC_DATA(n)                                                                      \
+	static struct eth_intel_igc_mac_tx tx_##n = {                                              \
+		.sem = tx_ring_lock_##n,                                                           \
+		.ring_wr_ptr = tx_ring_wr_ptr_##n,                                                 \
+		.ring_rd_ptr = tx_ring_rd_ptr_##n,                                                 \
+		.pkt = (struct net_pkt **)tx_pkt_##n,                                              \
+		.frag = (struct net_buf **)tx_frag_##n,                                            \
+	};                                                                                         \
+	static struct eth_intel_igc_mac_rx rx_##n = {                                              \
+		.sem = rx_ring_lock_##n,                                                           \
+		.ring_wr_ptr = rx_ring_wr_ptr_##n,                                                 \
+		.ring_rd_ptr = rx_ring_rd_ptr_##n,                                                 \
+	};                                                                                         \
+	static struct eth_intel_igc_mac_data eth_data_##n = {                                      \
+		.tx = &tx_##n,                                                                     \
+		.rx = &rx_##n,                                                                     \
```

Comment:
put struct eth_intel_igc_mac_tx and struct eth_intel_igc_mac_rx directly in struct eth_intel_igc_mac_data, not just as a pointer